Finally fix EOL-at-EOF and normalisation in ocamltest

dra27 opened the pull request:
In #8649, @damiendoligez noticed that ocamltest regards files containing `foo` and `foo\n` as identical - i.e. it ignores an LF at the end of the file.

This bug affects `typing-shadowing-of-pervasives-submodules/redefine_largefile.ml` test which has an EOL-at-EOF mark in the reference file, but doesn't actually produce one in the test.

While whiteboarding myself over this again, I noticed that there is another bug: at present ocamltest will not normalise CRLF to LF if the file doesn't have a CRLF at the end of the file. Parasitically, when normalising, ocamltest will incorrectly erase the final CR of a file.

So there are three bugs, 2 of which certainly matter. A reminder that CRLF normalisation means erasing a single CR character from before an LF iff every LF in a file is preceded by a CR character. In particular, a single CR at the end of the file should _not_ suppress normalisation of the rest of the file (it is as though the file ended CRCRLF). 

#8983 contains an implementation of comparison based on a function which returns a list of normalised lines. It is complex, as noted in https://github.com/ocaml/ocaml/pull/8983#pullrequestreview-353088235. On the basis that no one has yet proposed an alternative implementation which actually works, I would argue it is not _overly_ complex, but complex it certainly is.

I considered @shindere's suggestion in https://github.com/ocaml/ocaml/issues/8931#issue-491776897 to have a version of `input_line` which returns the terminator (or not if the terminator was in fact the EOF condition), @xavierleroy's comment in https://github.com/ocaml/ocaml/pull/8649#issuecomment-534652727 using sequences (which doesn't satisfy the normalising only if _all_ lines end CRLF), and @damiendoligez's comment about stream processing in https://github.com/ocaml/ocaml/pull/8983#issuecomment-583427325.

Now, you cannot write a function which returns a normalised list of lines to the specification above, because you must read the _entire_ file in order to know whether you are normalising or not. However, this made me consider if we are guilty of over-abstracting. Our final requirement is to know if two text files differ - this function to return the list of lines is supposed to be a means to that end, not the end itself.

If both files to be compared have CRLF line-endings, then there is no need to normalise them both - we open the files in binary mode and the `\r` will be present at the end of every line. The only time we actually need to normalise is if only _one_ file has CRLF line-endings (this is the common case on Windows - LF-formatted reference files coming from Git and CRLF endings coming from OCaml's printing functions). This observation, and dropping the reading a list of lines idea, allows what I think is the simplest implementation of a complex problem.

Using the definition that a line includes its LF terminator, the algorithm here is:
- Assume each file is CRLF-formatted if all the ignored lines and the first line of that file have a last character and it's `\r` (i.e. the opening lines up to the first which will be compared end `\r\n`)
- If only one file is CRLF-formatted, normalise its line and compare it to the other file's line. Otherwise no normalisation is ever performed.
- For each subsequent line, normalise if necessary and compare.
- If any line which should be normalised does not end with `\r\n`, **the files are different** and comparison stops.
- Otherwise the files are identical if the lines are identical.

This implementation adds the utility function `Seq.equal` to `Ocamltest_stdlib`. A sequence function is then defined which wraps `input_line` adding back in the `\n` (and correctly handling the `End_of_file` case). The function can also be instructed to strip `\r` characters and raises an exception if it can't.

`compare_text_files` now works by reading the "first" line of each file, determining which - if either - of the files should be normalised and setting up two of these sequences. The implementation completely replaces `Filecompare.compare_text_files`, so it's easier to review the actual file than the diff.

The implementation is not shorter than #8983 - it is, to me at least, somewhat more logical.

dra27 committed:
Fix redefine_largefile.ml test

The reference file included EOL at EOF but the test didn't produce one.

dra27 committed:
Fix handling of EOL-at-EOF in ocamltest

dra27 committed:
Report a diff for files differing by endings only

damiendoligez commented on ocamltest/filecompare.ml:
```diff
@@ -59,33 +59,103 @@ type files = {
+   the lines of [ic] where [first_line] is the already read first line of [ic].
+   Strings include the line terminator and CRLF is normalised to LF if
+   [normalise] is [true]. The sequences raises [Exit] if normalise is [true] and
+   a line is encountered which does not end CRLF. The final line of the file
+   only includes the terminator if it is presnt in the file (and a terminating
+   CR is never normalised if not strictly followed by LF). *)
```
```suggestion
   a terminated line is encountered which does not end CRLF. The final line
   of the file only includes the terminator if it is present in the file (and
   a terminating CR is never normalised if not strictly followed by LF). *)
```

dra27 replied:
Thanks - applied!

damiendoligez reviewed (approved):
Looks very good to me, and it's not that complex. Indeed that's the kind of thing I had in mind with my "stream processing" comment. Thanks for working out all the details.

dra27 commented:
Thanks for the review @damiendoligez - I've rebased, updated Changes and applied the clarification and correction to the comment in filecompare.ml. I'll merge once CI comes back.
